woocommerce/woocommerce-ios
# Analytics Hub: Fix range generation structure

Closes: #8615 

Why
==========
For stores with a vast order flow and numbers, it's possible to easily detect an inconsistency between the stats presented in the My Store section and those inside the Analytics Hub. This is due to a difference in how the My Store section builds the data request date range.

### My Store date range generation

Today: Jan 11, 2023
Time zone: GMT -3
Time: 11 AM GMT -3 or 14h UTC

    - Today
        - Start: 2023-01-11 03:00:00 UTC	
        - End: 2023-01-12 02:59:59 UTC	

    - This Week (named Week to Date in Analytics Hub)
        - Start: 2023-01-08 03:00:00 UTC	
        - End: 2023-01-15 02:59:59 UTC	

    - This Month (named Month to Date in Analytics Hub)
        - Start: 2023-01-01 03:00:00 UTC	
        - End: 2023-02-01 02:59:59 UTC	

    - This Year ((named Year to Date in Analytics Hub))
        - Start: 2023-01-01 03:00:00 UTC	
        - End: 2024-01-01 02:59:59 UTC	

### Current Analytics Hub date range generation
Today: Jan 11, 2023
Time zone: GMT -3
Time: 11 AM GMT -3 or 14h UTC

    - Today
        - Start: 2023-01-11 03:00:00 UTC
        - End: 2023-01-11 14:00:00 UTC

    - Week to Date
        - Start: 2023-01-08 03:00:00 UTC
        - End: 2023-01-11 14:00:00 UTC

    - Month to Date
        - Start: 2023-01-01 03:00:00 UTC 
        - End: 2023-01-11 14:00:00 UTC

    - Year to Date
        - Start: 2023-01-01 03:00:00 UTC
        - End: 2023-01-11 14:00:00 UTC

The My Store section creates the end date far in the future for each tab instead of using the current time. This covers any time zone gap between the app and the store, always fetching as much data in “the future” as possible.

But inside the Analytics Hub, the `to Date` selections will use the end date of the range as the exact instant where the data was requested. This difference between the My Store and the Analytics Hub date range end is the issue source since the Time zone difference between the app and the store will cause the Hub not to fetch everything, while the My Store will.

How
==========
Adjust the internal range generation classes inside the Analytics Hub to follow the same My Store date range generation rule while keeping the expected description. To achieve this, it was necessary to generate the formatted string of each range now range inside each data class, making the request data the same as the `My Store` one for the `Today`, `Week to Date`, `Month to Date`, `Quarter to Date` and `Year to Date`, while keeping the UI description of the ranges the same.

Capture
==========
### Before
https://user-images.githubusercontent.com/5920403/211960994-a7caec68-d0e3-4c68-9dcd-8a2efa5f2fdf.mp4

### After
https://user-images.githubusercontent.com/5920403/211962894-66e4fd99-1429-43b0-8529-71aa63e22dd9.mp4

How to Test
==========
1. Navigate between the My Store four tabs and the respective Analytics Hub screen and verify there's no difference in the numbers between both screens.

PS: The Flying Origami store is highly recommended for reproducing this test scenario.
